src/ilasm/assem.cpp

```diff
@@ -823,17 +823,21 @@ BOOL Assembler::EmitMethod(Method *pMethod)
     {
         ULONG i;
         mdToken* ptk;
-        mdToken tk;
+        mdToken tk = 0;
         for(i = 0; i < pMethod->m_NumTyPars; i++)
         {
-            //ptk = (pMethod->m_TyParBounds[i] == NULL)? NULL :  (mdToken*)(pMethod->m_TyParBounds[i]->ptr());
-            //if(FAILED(m_pEmitter->DefineGenericParam(MethodToken,i,0,pMethod->m_TyParNames[i],0,ptk,&tk)))
             ptk = (pMethod->m_TyPars[i].Bounds() == NULL)? NULL :  (mdToken*)(pMethod->m_TyPars[i].Bounds()->ptr());
-            if(FAILED(m_pEmitter->DefineGenericParam(MethodToken,i,pMethod->m_TyPars[i].Attrs(),pMethod->m_TyPars[i].Name(),0,ptk,&tk)))
-                report->error("Unable to define generic param'\n");
+            if (FAILED(m_pEmitter->DefineGenericParam(MethodToken, i, pMethod->m_TyPars[i].Attrs(), pMethod->m_TyPars[i].Name(), 0, ptk, &tk)))
```

Comment:
It feels like there may be an issue here where the constraint list calculated for DefineGenericParam and the constraint list calculated later in EmitGenericParamConstraints may get out of sync. I would like to see a few cases where the .param constraint directives are a subset/superset of the constraints described as the type parameters are defined.

Reply:
This code handles the definition of Generic Parameters:  (i,e T, or T1 and T2)
and any associated Custom attributes, recorded in `pMethod->m_TyPars[i].CAList()`

After these are emitted we then emit a separate list of  Generic Parameter Constraints
held in `pMethod->m_GPCList`, which are tuples of `< m_iGenericParamIndex, m_tkTypeConstraint >`

This is the code for Generic Methods,  and we have the same code for Generic Classes just below here (line 1236)

Reply:
We have to sequence everything that we emit, because we have to create the tokens that we need, (i,e the GenericParameter, that we record here on line 836) before we can create the GenericParamConstraint tokens.
Only after we have created this GenericParamConstraint can we emit the Custom attributes that we attached to 
a m_GPCList.

Most of this work takes place in `EmitGenericParamConstraints()`

Reply:
This has been fixed with the recent updates. Please test that this works, but the code looks good to me now.

---
In reply to: [315328900](https://github.com/dotnet/coreclr/pull/26219#discussion_r315328900) [](ancestors = 315328900)